js/ccf-app/DEVELOPMENT.md

```diff
@@ -0,0 +1,78 @@
+# Development notes
+
+NOTE: Before contributing to this package, please read this document in full.
+
+## Getting started
+
+```sh
+npm install
+```
+
+## Code
+
+**Relative imports**:
+Every relative import of another TypeScript module **must** end with `.js` (yes, not `.ts`).
+The extension is the **target** extension and all TypeScript-related tooling understands this.
+Omitting extensions will lead to omitted extensions in the transpiled `.js` files
+which may cause import issues depending on the module loader used in the consuming environment.
+
+## Tests
+
+```sh
+npm test
+```
+
+Tests make use of the `src/polyfill.ts` module and run in Node.js, not in CCF.
+This allows to iterate quickly and allows to test most functions.
+
+End-to-end tests running inside CCF are located in `/tests/npm-app` and are run as part of CCF's regular CI.
+
+## Debugging
+
+Tests executed via `npm test` run in Node.js and can be debugged easily.
+The simplest workflow is using the [JavaScript Debug Terminal](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_javascript-debug-terminal) in VS Code. Set a break point and run `npm test` inside the JavaScript Debug Terminal. The debugger will attach and stop automatically.
+
+Debugging problems that only happen in CCF is harder.
+
+If the problem occurs in native functions/properties exposed from C++ (see `src/global.ts`) then VS Code's C++ debugger can be used when running a CCF end-to-end test in virtual mode (not SGX).
+
+Debugging of JavaScript code running in CCF is currently not possible.
+A compromise is to run a debug build of CCF with verbose logging enabled.
+When running JavaScript code, uncaught exceptions and output from `console.log()` is dumped to the node's log files.
```

Comment:
Why is verbose logging recommended? `console.log` is `INFO` level, so they can `console.log` printf-debug their own apps without verbose logging?

Reply:
I see, I wasn't aware of that. What about uncaught exceptions? Are they also INFO or higher?